File: packages/cli/src/commands/agent-profile.ts
import { Command } from "commander";
import {
  buildAgentProfile,
  defaultAgentProfilePath,
  readAgentProfile,
  writeAgentProfile,
} from "../core/agent-profile.js";
import { CliError } from "../core/exit.js";

interface AgentProfileOptions {
  agent?: string;
  fromBenchmark?: string;
  out?: string;
  profile?: string;
  root?: string;
  json?: boolean;
}

export function agentProfileCommand(): Command {
  const cmd = new Command("agent-profile").description(
    "Build advisory agent capability profiles from benchmark reports"
  );

  cmd
    .command("update")
    .description("Create or update an advisory profile from benchmark output")
    .requiredOption("--agent <id>", "Agent id")
    .option("--from-benchmark <path>", "Benchmark report JSON/YAML path")
    .option("--out <path>", "Profile output path")
    .option("--root <path>", "Repository root", process.cwd())
    .option("--json", "Output JSON instead of text", false)
    .action(async (opts: AgentProfileOptions) => {
      const profile = await buildAgentProfile({
        agentId: opts.agent as string,
        benchmarkReportPath: opts.fromBenchmark,
      });
      const outPath = await writeAgentProfile(
        profile,
        opts.out ??
          defaultAgentProfilePath(opts.root ?? process.cwd(), profile.agent_id)
      );
      const output = { ok: true, path: outPath, profile };
      if (opts.json) console.log(JSON.stringify(output, null, 2));
      else {
        console.log(`# x-harness Agent Profile: ${profile.agent_id}`);
        console.log(
          `- observed_failure_modes: ${profile.observed_failure_modes.length}`
        );
        console.log(
          `- required_extra_checks: ${profile.required_extra_checks.join(", ")}`
        );
        if (outPath) console.log(`- path: ${outPath}`);
      }
    });

  cmd
    .command("report")
    .description("Read and validate an advisory agent profile")
    .option("--profile <path>", "Profile JSON/YAML path")
    .option("--agent <id>", "Agent id for the default profile path")
    .option("--root <path>", "Repository root", process.cwd())
    .option("--json", "Output JSON instead of text", false)
    .action(async (opts: AgentProfileOptions) => {
      const profilePath =
        opts.profile ??
        (opts.agent
          ? defaultAgentProfilePath(opts.root ?? process.cwd(), opts.agent)
          : null);
      if (!profilePath) {
        throw new CliError(
          "agent-profile report requires --profile or --agent",
          2
        );
      }
      const profile = await readAgentProfile(profilePath);
      if (opts.json) console.log(JSON.stringify(profile, null, 2));
      else {
        console.log(`# x-harness Agent Profile: ${profile.agent_id}`);
        console.log(`- advisory_only: ${profile.advisory_only}`);
        console.log(`- admission_authority: ${profile.admission_authority}`);
      }
    });

  return cmd;
}

File: internal/cli/agentprofile.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"path/filepath"
	"strings"

	"github.com/BrianNguyen29/x-harness/internal/agentprofile"
)

func handleAgentProfile(args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, "agent-profile requires a subcommand: update, report")
		return ExitUsage
	}

	switch args[0] {
	case "update":
		return handleAgentProfileUpdate(args[1:], stdout, stderr)
	case "report":
		return handleAgentProfileReport(args[1:], stdout, stderr)
	default:
		fmt.Fprintf(stderr, "unknown agent-profile subcommand: %s\n", args[0])
		return ExitUsage
	}
}

func handleAgentProfileUpdate(args []string, stdout, stderr io.Writer) int {
	var agentID, benchmarkPath, outPath, root string
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--agent":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --agent requires a value")
				return ExitUsage
			}
			agentID = args[i+1]
			i++
		case "--from-benchmark":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --from-benchmark requires a value")
				return ExitUsage
			}
			benchmarkPath = args[i+1]
			i++
		case "--out":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --out requires a value")
				return ExitUsage
			}
			outPath = args[i+1]
			i++
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return ExitUsage
			}
			root = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if agentID == "" {
		fmt.Fprintln(stderr, "Error: --agent <id> is required")
		return ExitUsage
	}

	if root == "" {
		root = "."
	}
	root, _ = filepath.Abs(root)

	profile, err := agentprofile.BuildAgentProfile(agentID, benchmarkPath)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if err := agentprofile.ValidateAgentProfile(profile, root); err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if outPath == "" {
		outPath = agentprofile.DefaultAgentProfilePath(root, agentID)
	}

	if err := agentprofile.WriteAgentProfile(profile, outPath); err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		output := map[string]any{
			"ok":      true,
			"path":    outPath,
			"profile": profile,
		}
		data, _ := json.MarshalIndent(output, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintf(stdout, "# x-harness Agent Profile: %s\n", profile.AgentID)
		fmt.Fprintf(stdout, "- observed_failure_modes: %d\n", len(profile.ObservedFailureModes))
		fmt.Fprintf(stdout, "- required_extra_checks: %s\n", strings.Join(profile.RequiredExtraChecks, ", "))
		fmt.Fprintf(stdout, "- path: %s\n", outPath)
	}

	return ExitOK
}

func handleAgentProfileReport(args []string, stdout, stderr io.Writer) int {
	var profilePath, agentID, root string
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--profile":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --profile requires a value")
				return ExitUsage
			}
			profilePath = args[i+1]
			i++
		case "--agent":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --agent requires a value")
				return ExitUsage
			}
			agentID = args[i+1]
			i++
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return ExitUsage
			}
			root = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if root == "" {
		root = "."
	}
	root, _ = filepath.Abs(root)

	if profilePath == "" {
		if agentID == "" {
			fmt.Fprintln(stderr, "Error: agent-profile report requires --profile or --agent")
			return ExitUsage
		}
		profilePath = agentprofile.DefaultAgentProfilePath(root, agentID)
	}

	profile, err := agentprofile.ReadAgentProfile(profilePath)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if err := agentprofile.ValidateAgentProfile(profile, root); err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		data, _ := json.MarshalIndent(profile, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintf(stdout, "# x-harness Agent Profile: %s\n", profile.AgentID)
		fmt.Fprintf(stdout, "- advisory_only: %v\n", profile.AdvisoryOnly)
		fmt.Fprintf(stdout, "- admission_authority: %v\n", profile.AdmissionAuthority)
	}

	return ExitOK
}
